### PR TITLE
Add server side code to create inactive users csv file for AB#12015.

### PR DESCRIPTION
**Pull request by BrianMaki**
# Fixes or Implements [AB#12015](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/12015)

## Description

Created new HttpGet call for Admin in CvsExportController with route called 'GetInactiveUsers' and produces 'text/csv'.

Service will retrieve a list of inactive users from the database based on a days inactive parameter. The service will then further filter that list by only including inactive users whose roles are in AdminUser and/or SupportUser roles.  This information is stored in Keycloak.

Created or updated following:

1. Add call to CvsExportController
2. Add call to CvcExportService
3. Add call to DbAdminUserProfileDelegate
4. Add call to KeycloakUserAdminDelegate
5. Add code to generate csv file
6. Added tests to KeycloakUserAdminDelegateTests
7. Removed mock configuration call in AuthenticationServiceTests
8. Updated all environment appsettings.json with Keycloak admin configuration values
9. Addressed some spelling issues in documenation.

Following example of csv generated. Opened using Excel.

Screen Shot 2022-01-08 at 4 28 34 PM

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

-   [x] Unit Tests Updated
-   [ ] Functional Tests Updated
-   [ ] Not Required

### UI Changes

NO

### Browsers Tested

-   [ ] Chrome
-   [ ] Safari
-   [ ] Edge
-   [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant. Include any specialized deployment steps here.

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

-   Fulfills Work Item Requirements

    -   The changes meet/implement story's acceptance criteria. Fixes identified problems if bug.
    -   Changes not directly related to the task requirements need to be documented on the PR.

-   Compilation / Tests

    -   The changes do not break compilation and/or unit or functional tests; unless other item addresses them specifically.

-   Logic Problems / Functional Behaviour

    -   The changes work as intended and do not introduce problems.

-   Performance

    -   The changes do not introduce obvious performance issues.

-   Documented Standards

    -   The changes adhere to the team's documented [Coding Standards](https://github.com/bcgov/healthgateway/wiki/standards).

-   Readability / Maintainability
    -   The changes can be easily understood/read and allow for future enhancements without major refactoring. Readability is preferred over "clever code".
    -   Reasoning for changes needs to be clearly articulated. Disagreements should be arbitrated by a third developer.

### CODE REVIEW

File: Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs [hunk 1]
```diff
@@ -85,6 +99,51 @@ public UserRepresentation GetUser(Guid userId, JWTModel jwtModel)
...
+                    retVal.TotalResultCount = users.ToList().Count;
+                }
+            }
+            else
+            {
+                this.logger.LogError("Unable to connect to endpoint: {Endpoint}, HTTP Error: {Response} ...", baseUri + uri, response.StatusCode);
+
+                retVal.ResultError = new RequestResultError()
+                {
+                    ResultMessage = $"Unable to connect to endpoint: {baseUri}{uri}, HTTP Error: {response.StatusCode}.",
```
Comment (sslaws) [inline Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs:139]: Why not generate this string once and then you can log the same string on 135 above

**Review by sslaws — DISMISSED** [review]

File: Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs (hunk 1 above)
Comment (BrianMaki) [reply Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs:139]: I wanted to use a message defined in a string for both ResultError and Logger but I got a warning from the logger.  The logger gave a warning when the message was created before hand.

File: Apps/AdminWebClient/src/Server/Services/CsvExportService.cs [hunk 2]
```diff
@@ -83,6 +122,44 @@ public Stream GetRatings(DateTime? startDate, DateTime? endDate)
...
+        public async Task<Stream> GetInactiveUsers(int inactiveDays)
+        {
+            this.logger.LogDebug("Getting inactive users since {InactiveDays} day(s) from last login....", inactiveDays);
+
+            List<AdminUserProfileView> inactiveUsers = new List<AdminUserProfileView>();
+
+            // Inactive admin user profiles from DB
+            DBResult<IEnumerable<AdminUserProfile>> profilesResult = this.adminUserProfileDelegate.GetInactiveAdminUserProfiles(inactiveDays);
+
+            if (profilesResult.Status == DBStatusCode.Read && profilesResult.Payload.ToList().Count > 0)
```
Comment (MikeLyttle) [inline Apps/AdminWebClient/src/Server/Services/CsvExportService.cs:135]: Nit: `profilesResult.Payload.ToList().Count` can be replaced with `profilesResult.Payload.Count()` which should avoid creating an extra List object.

**Review by pagustinqt20 — APPROVED** [review]

**Review by MikeLyttle — DISMISSED** [review]

File: Apps/Database/src/Delegates/DbAdminUserProfileDelegate.cs [hunk 3]
```diff
@@ -67,6 +69,24 @@ public DBResult<AdminUserProfile> GetAdminUserProfile(string username)
...
+        this.logger.LogTrace("Retrieving all the inactive admin user profiles for the past {InactiveDays} day(s)...", inactiveDays);
+
+        DBResult<IEnumerable<AdminUserProfile>> result = new DBResult<IEnumerable<AdminUserProfile>>()
+        {
+            Payload = this.dbContext.AdminUserProfile
+                .Where(profile => profile.LastLoginDateTime.Date <= DateTime.UtcNow.AddDays(-inactiveDays).Date)
+                .OrderByDescending(profile => profile.LastLoginDateTime)
+                .ToList(),
+        };
+        result.Status = result.Payload != null ? DBStatusCode.Read : DBStatusCode.NotFound;
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/Database/src/Delegates/DbAdminUserProfileDelegate.cs]: Addressed.

File: Apps/Database/src/Delegates/DbAdminUserProfileDelegate.cs [hunk 4]
```diff
@@ -67,6 +69,24 @@ public DBResult<AdminUserProfile> GetAdminUserProfile(string username)
...
+
+        this.logger.LogTrace("Finished retrieving {Count} inactive admin user profiles for the past {InactiveDays} day(s)...", result.Payload!.ToList().Count, inactiveDays);
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/Database/src/Delegates/DbAdminUserProfileDelegate.cs]: Addressed.

File: Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs [hunk 5]
```diff
@@ -85,6 +99,51 @@ public UserRepresentation GetUser(Guid userId, JWTModel jwtModel)
...
+
+            HttpResponseMessage response = await client.GetAsync(new Uri(uri, UriKind.Relative)).ConfigureAwait(true);
+            string payload = await response.Content.ReadAsStringAsync().ConfigureAwait(true);
+            this.logger.LogDebug("Getting users for role: {Role}, response payload: {Payload} ...", role, payload);
+
+            if (response.IsSuccessStatusCode)
+            {
+                retVal.ResultStatus = ResultType.Success;
+
+                List<UserRepresentation>? users = JsonSerializer.Deserialize<List<UserRepresentation>>(payload);
```
[the remark this answers is not in the document]
Comment (BrianMaki) [reply Apps/Common/src/AccessManagement/Administration/KeycloakUserAdminDelegate.cs]: Added try/catch exception - same as RestLaboratoryDelegate

File: Apps/AdminWebClient/src/Server/Services/CsvExportService.cs (hunk 2 above)
Comment (BrianMaki) [reply Apps/AdminWebClient/src/Server/Services/CsvExportService.cs]: Addressed. Used *.Any()

**Review by MikeLyttle — DISMISSED** [review]

**Review by MikeLyttle — APPROVED** [review]

**Review by pagustinqt20 — APPROVED** [review]

**Pull request merged**